universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java

```diff
@@ -189,23 +189,38 @@ public ProgramDefinition moveBlock(long blockId, Direction direction)
         blockSlice.startsBefore(otherBlockSlice) ? blockSlice : otherBlockSlice;
     BlockSlice latterSlice =
         blockSlice.startsBefore(otherBlockSlice) ? otherBlockSlice : blockSlice;
+
+    // An illegal move is when a predicate in a later block depends on an earlier block.
+    // Get the questions used in any predicate in the latter block.
+    ImmutableSet<Long> predicateQuestions =
+        blockDefinitions().subList(latterSlice.startIndex(), latterSlice.endIndex()).stream()
+            .filter(b -> b.visibilityPredicate().isPresent())
+            .flatMap(b -> b.visibilityPredicate().get().getQuestions().stream())
+            .collect(toImmutableSet());
+
+    if (!predicateQuestions.isEmpty()) {
+      if (blockDefinitions().subList(earlierSlice.startIndex(), earlierSlice.endIndex()).stream()
+          .flatMap(b -> b.programQuestionDefinitions().stream().map(ProgramQuestionDefinition::id))
+          .anyMatch(predicateQuestions::contains)) {
+        throw new IllegalBlockMoveException(
+            "This move is not possible - it would move a block condition before the question it"
+                + " depends on");
+      }
+    }
+
     ImmutableList.Builder<BlockDefinition> blocksBuilder = ImmutableList.builder();
 
     // Swap the two slices, assuming these slices are consecutive slices
-    blockDefinitions().stream()
-        .limit(earlierSlice.startIndex())
-        .forEach(blockDefinition -> blocksBuilder.add(blockDefinition));
+    blockDefinitions().stream().limit(earlierSlice.startIndex()).forEach(blocksBuilder::add);
     blockDefinitions().stream()
         .skip(latterSlice.startIndex())
         .limit(latterSlice.endIndex() - latterSlice.startIndex())
-        .forEach(blockDefinition -> blocksBuilder.add(blockDefinition));
+        .forEach(blocksBuilder::add);
```

Comment:
praise: nice!

Reply:
I just clicked the IntelliJ idea button, I take no credit :)